### PR TITLE
Make creation and visitation of SyntaxData more efficient

### PR DESCRIPTION
**Pull request by akyrtzi**
* Make `SyntaxData` an efficient value type that eagerly tracks how a raw node relates to a source file tree, like its absolute offset and its parent.
* `SyntaxData` has no identity and Syntax types are no longer `Equatable`/`Hashable`.
* `Syntax` objects are fully thread-safe now, `LazyNonThreadSafeCache` has been removed.
* `child(at:)` and `numberOfChildren` have been removed from the Syntax API due to `child(at:)`'s performance becoming `O(n)`. Clients should be using the provided `Sequence`s for efficient visitation.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb [hunk 1]
```diff
@@ -66,8 +61,7 @@ public protocol ${node.name}: Syntax {}
 %     for line in dedented_lines(node.description):
 /// ${line}
 %     end
-/// Syntax nodes are not thread-safe.
-public struct ${node.name}: ${base_type}, _SyntaxBase, Hashable {
```
Comment (harlanhaskins) [inline Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: I think we don't need to give up `Equatable`/`Hashable` with this patch. Could we use the absolute position from the beginning of the file as the identity?

```swift
public static func ==(lhs: Syntax, rhs: Syntax) -> Bool {
  return base.position
}

public func hash(into hasher: inout Hasher) {
  hasher.combine(base.position)
}
```

This exposes an invariant that you can't rely on `==`/`hashValue` between trees, which is unfortunate. But it's par for the source with SwiftSyntax -- nodes belonging to different trees have always broken assumptions about the questions you can ask. For example, calling `.sourceLocation` for two nodes from two different trees will give you views into wildly different contexts, and that's no different here.

File: Sources/SwiftSyntax/Syntax.swift [hunk 2]
```diff
@@ -196,105 +161,89 @@ extension Syntax {
...
   /// leading trivia attached to the first token syntax.
-  public var positionAfterSkippingLeadingTrivia: AbsolutePosition {
+  var positionAfterSkippingLeadingTrivia: AbsolutePosition {
     return data.positionAfterSkippingLeadingTrivia
   }
 
   /// The absolute position where this node (excluding its trailing trivia)
   /// ends.
-  public var endPosition: AbsolutePosition {
+  var endPosition: AbsolutePosition {
```
Comment (harlanhaskins) [inline Sources/SwiftSyntax/Syntax.swift]: I feel like, if these are going to become internal, we might as well get rid of them and just dispatch through `data` inside the module. What do you think?

File: Sources/SwiftSyntax/Syntax.swift [hunk 3]
```diff
@@ -320,22 +269,219 @@ extension Syntax {
...
+    return base.byteSizeAfterTrimmingTrivia
+  }
+
+  /// The root of the tree in which this node resides.
+  public var root: Syntax {
+    return base.root
+  }
+
+  /// Passes to a closure every present token node that is part of this node.
+  public func forEachToken(_ receiver: (TokenSyntax)->()) {
```
Comment (harlanhaskins) [inline Sources/SwiftSyntax/Syntax.swift]: I should have raised this concern earlier, but I'm wary of adding `forEach`-style methods without an equivalent `Sequence`-type. Could we, instead of exposing `forEachToken`, add a `TokenSequence` type that performs the same duty? That gives callers a lot more freedom, and more importantly, allows them to compose with operations like `map()`.

```
for token in node.tokens {
  // we don't need to capture in the closure here, which could save allocations
}

let text = node.tokens.map { $0.text }.joined()
```

File: Sources/SwiftSyntax/Syntax.swift (hunk 2 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: I assume you mean all four position-related vars. I don't have a particular objection but could you clarify what is the benefit of removing them ?

Comment (harlanhaskins) [reply Sources/SwiftSyntax/Syntax.swift]: Just because they don't serve much purpose, and they're an extra hop. They'll likely be inlined anyway, but still.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb (hunk 1 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: Oooh I have plenty of thoughts about this, I'd love to get your feedback 🙂.

The absolute position can work sufficiently if you check for identity of `TokenSyntax`s since a present TokenSyntax is guaranteed to have a unique position. For example it would work for swift-format's only use-case of [using Syntax as Hashable](https://github.com/google/swift/blob/abb642710435be7f12a0e404dfaef1ac5f3d068c/Sources/SwiftFormatPrettyPrint/TokenStreamCreator.swift#L21).
The other use I saw for Syntax as dictionary key was with SwiftSyntax's own `SyntaxClassifier`, whose use of a Dictionary did not seem important for its functionality and I changed it in https://github.com/apple/swift-syntax/pull/60. (I also intend to change it further to a simpler sequence but this is another PR).

However, position does not provide uniqueness in other nodes in the tree since a parent node has the same position as its first child. You could maybe combine the position with the syntax kind but it's not clear to me whether a `(position, syntax kind)` tuple guarantees uniqueness or not.

An alternative I considered was that we could have an `indexInTree: Int` property that would represent the unique index of the node in its root tree. This could be done efficiently by:
* Having `RawSyntax` store not just its `SourceLength` but also how many total nodes it represents.
* Add `indexInTree` in `AbsoluteSyntaxInfo` and update it appropriately while walking the tree:
```
struct AbsoluteSyntaxInfo {
  let offset: UInt32
  let indexInParent: UInt32
  let indexInTree: UInt32
}
```
(This would make `AbsoluteSyntaxInfo` larger than a `UInt64` but we could also consider not bothering to keep track of `indexInParent`, which is only useful for `replaceSelfWith` modifications.)

Now, to clarify a bit why I tentantively chose to remove the `Equatable/Hashable` conformance:
* It's not clear to me how important it is to provide a unique identity for each node. For `TokenSyntax` the position is sufficient and adapting for swift-format is simple. Perhaps we could make this change and see in practice if there are compelling use cases for SwiftSyntax clients for a unique node identity. If it proves important we can look into adding `indexInParent`.
* Even with having an `indexInParent` property available I have concerns with presenting Syntax nodes as `Equatable/Hashable`. It seems to me it would confuse people over what a Syntax node represents, which is a view of a node in a particular tree configuration and comparing nodes *directly* is not sensible unless they came from the same tree. I feel that using a node's position, or indexInTree, instead makes things much more clear. For example, if I see code doing:
```
if node1.position == node2.position {...}
```
or
```
if node1.indexInTree == node2.indexInTree {...}
```
it is obvious in the code what are the expectations of that comparison, they are properties describing how a node relates in a specific tree. There may even be use cases to compare positions from different trees.

Otherwise with
```
if node1 == node2 {...}
```
The semantics of this kind of comparison are completely non-obvious. What can I infer from this about what it means to compare nodes from different trees ? I think the intuitive assumption with such a code would be that each node is unique and every node from a tree would compare differently from another tree, which is not the case now; it may lead people to make incorrect use of the Syntax nodes.

Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: > future extensions of syntax tree such as semantic model, where we associate more information with a syntax node as side table

Let's say we have syntax nodes associated with semantic info; then when you make any change to a syntax tree it is very likely that other syntax nodes would have their semantic information invalidated. I don't think it is correct to assume that because a syntax node had some semantic info associated with in a particular tree, that you can just keep the same semantic info for that syntax node in a different tree configuration.

In other words, I think this particular use case would not benefit from unique identity across trees.

Comment (nkcsgexi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: I agree for clients like swift-formatter, there're no need to differentiate `this exact syntax node` and `syntax node of this kind in general`. But once we start to provide more context-sensitive info in the swift-syntax APIs, the uniqueness of each syntax node starts to become necessary IMO.

Comment (nkcsgexi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: > Let's say we have syntax nodes associated with semantic info; then when you make any change to a syntax tree it is very likely that other syntax nodes would have their semantic information invalidated.

That's a good point. But it's equally likely that some nodes' semantic information stays stable so we can re-use it without performing unnecessary work. Whether using cache or re-calculating the semantic information of a node should be implementation detail of the semantic model. But it seems to me we'll always have to recalculate semantic information without being hashable.

Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: > Having a unique identifier for each node

To clarify a bit more here, the previous model of Syntax nodes was that every node was unique, and every tree had nodes unique from any other tree. You could not use the node identity to transfer information from one tree to another, all the nodes of the new tree would have completely different identities. Even the raw nodes that got reused during reparsing ended up wrapped by whole new nodes with different identities.

To transfer info from one tree to another you would have to use another property, like perhaps source ranges, unique node identity is not helping you.

Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: Here are some more thoughts. If we want unique node identify across trees here's what we can do:
* Add `indexInTree`
* Add an atomically incremented counter for every root node that gets created.
Then we have something like:
```
struct AbsoluteSyntaxInfo {
  let offset: UInt32
  let nodeId: NodeIdentity
}

struct NodeIdentity {
  let rootId: UInt32
  let indexInTree: UInt32
}
```
`rootId` is unique for each root across the process, `indexInTree` is unique within a root tree. With a unique node identity across trees, making nodes `Equatable/Hashable` would be sensible I think.

File: Sources/SwiftSyntax/Syntax.swift (hunk 2 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: I realized a subtle problem with removing properties from `_SyntaxBase` and only have them on `Syntax`.
The point of mirroring the `Syntax` API over `_SyntaxBase` is so that internally we can call the APIs and avoid the unnecessary checks that the object conforms to `_SyntaxBase`.
But if an API exists only on `Syntax` and not on `_SyntaxBase` then it is easy to inadvertently call it internally. For example if there is code such as `node.position` where `node` is a `_SyntaxBase`, by removing `position` from `_SyntaxBase` it makes the call go through `Syntax` and the unnecessary check.

File: Sources/SwiftSyntax/Syntax.swift [hunk 4]
```diff
@@ -320,22 +269,219 @@ extension Syntax {
...
+  }
+
+  /// A sequence over the `present` children of this node.
+  public var children: SyntaxChildren {
+    return SyntaxChildren(base)
+  }
+
+  /// A reversed sequence over the `present` children of this node.
+  /// This is more efficient than doing `children.reversed()`.
+  public var childrenReversed: SyntaxChildrenReversed {
```
[the remark this answers is not in the document]
Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: Great idea! Done.

File: Sources/SwiftSyntax/Syntax.swift (hunk 3 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: Good point, much better! Done.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb (hunk 1 above)
Comment (harlanhaskins) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: That could definitely work. It's unfortunate that it's no longer a word, but I think having this, and re-enabling `Equatable`/`Hashable` would be great in the long run.

File: Tests/SwiftSyntaxTest/SyntaxTests.swift [hunk 5]
```diff
@@ -23,7 +23,9 @@ public class SyntaxAPITestCase: XCTestCase {
     XCTAssertEqual("\(funcKW.previousToken!)", "{ ")
 
     var toks = [TokenSyntax]()
-    funcKW.parent!.forEachToken { toks.append($0) }
+    for token in funcKW.parent!.tokens {
```
Comment (harlanhaskins) [inline Tests/SwiftSyntaxTest/SyntaxTests.swift]: Nitpick (not essential, but a good outcome of switching to sequences):

```swift
let toks = Array(funkKW.parent!.tokens)
```

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb (hunk 1 above)
Comment (nkcsgexi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: > * Add `indexInTree`
> * Add an atomically incremented counter for every root node that gets created.
>   Then we have something like:
This sounds a great approach for me!

File: Tests/SwiftSyntaxTest/SyntaxTests.swift (hunk 5 above)
Comment (akyrtzi) [reply Tests/SwiftSyntaxTest/SyntaxTests.swift]: Nice! Done.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb (hunk 1 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: Please take a look into the latest commit.

File: Sources/SwiftSyntax/Syntax.swift [hunk 6]
```diff
@@ -454,10 +454,20 @@ extension Syntax {
   }
 }
 
+/// Determines if two nodes are equal to each other.
+func ==(lhs: _SyntaxBase, rhs: _SyntaxBase) -> Bool {
+  return lhs.data.nodeId == rhs.data.nodeId
+}
+
+/// Determines if two nodes are equal to each other.
+public func ==(lhs: Syntax, rhs: Syntax) -> Bool {
```
Comment (harlanhaskins) [inline Sources/SwiftSyntax/Syntax.swift]: These seem left-over from testing.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb (hunk 1 above)
Comment (harlanhaskins) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb]: Thank you, @akyrtzi!

File: Sources/SwiftSyntax/Syntax.swift (hunk 6 above)
Comment (harlanhaskins) [reply Sources/SwiftSyntax/Syntax.swift]: Oh wait, no, this is to work around the fact that `Syntax` itself is not Equatable. Which is the right move.

Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: On a related note, should we expose `nodeId` from the `Syntax` API ? Otherwise you can't use a bare `Syntax` node as a dictionary key, you have to cast it to a concrete type.

Comment (akyrtzi) [reply Sources/SwiftSyntax/Syntax.swift]: > you have to cast it to a concrete type

To clarify, that was the case before any of these changes.

Comment (harlanhaskins) [reply Sources/SwiftSyntax/Syntax.swift]: I don't think it would hurt. Let's make sure the properties are `internal`, and maybe rename it to something like `identifier` or `uniqueIdentifier`.

File: Sources/SwiftSyntax/SyntaxData.swift [hunk 7]
```diff
@@ -49,7 +49,7 @@ struct AbsoluteSyntaxInfo {
 }
 
 /// Provides a stable and unique identity for `Syntax` nodes.
-struct NodeIdentity: Hashable {
+public struct NodeIdentity: Hashable {
```
Comment (harlanhaskins) [inline Sources/SwiftSyntax/SyntaxData.swift]: Apologies for the back-and-forth on this -- can you rename this to `SyntaxIdentifier` so it looks more like something from the `Syntax` namespace? We don't often use `node` in API.

Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxData.swift]: SGTM, done!

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by harlanhaskins** [comment]
This looks great! :boat:
